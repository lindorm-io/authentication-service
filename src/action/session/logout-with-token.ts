import Joi from "@hapi/joi";
import { JOI_JWT_TOKEN } from "../../constant";
import { IAuthContext } from "../../typing";
import { assertAccountPermission } from "../../support";

export interface ILogoutWithTokenOptions {
  refreshToken?: string;
}

const schema = Joi.object({
  refreshToken: JOI_JWT_TOKEN,
});

export const logoutWithToken = (ctx: IAuthContext) => async (options?: ILogoutWithTokenOptions): Promise<void> => {
  await schema.validateAsync(options);

  const { repository, token } = ctx;
  const {
    refresh: { subject: sessionId },
  } = token;

  const session = await repository.session.find({ id: sessionId });

  await assertAccountPermission(ctx)(session.accountId);

  await repository.session.remove(session);
};
