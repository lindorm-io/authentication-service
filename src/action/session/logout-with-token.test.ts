import { logoutWithToken } from "./logout-with-token";
import {
  getTestRepository,
  getTestAccount,
  getTestClient,
  getTestSession,
  inMemoryStore,
  resetStore,
} from "../../test";
import { baseHash } from "@lindorm-io/core";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("../../support", () => ({
  assertAccountPermission: jest.fn(() => () => {}),
  encryptClientSecret: jest.fn((input) => baseHash(input)),
}));

describe("logoutWithToken", () => {
  let ctx: any;

  beforeEach(async () => {
    ctx = {
      repository: await getTestRepository(),
      token: { refresh: { subject: "be3a62d1-24a0-401c-96dd-3aff95356811" } },
    };
    await ctx.repository.session.create(
      getTestSession(
        getTestAccount("email@lindorm.io"),
        getTestClient(),
        "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
        "sha256",
      ),
    );
  });

  afterEach(resetStore);

  test("should log out", async () => {
    await expect(logoutWithToken(ctx)({ refreshToken: "token.token.token" })).resolves.toBe(undefined);
    expect(inMemoryStore).toMatchSnapshot();
  });
});
