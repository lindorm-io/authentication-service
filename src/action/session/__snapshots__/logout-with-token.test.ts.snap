// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logoutWithToken should log out 1`] = `
Object {
  "authentication": Object {
    "session": Object {
      "data": Array [],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;
